### PR TITLE
Support for external storage workloads

### PR DESCRIPTION
**Pull request by manik-couchbase**
- Support for external storage
- Currently only supports AWS S3
- Operations supported : [Create/Delete Bucket, Create/Delete Folders, Insert / Update / Delete Files, Insert Files into Folders, Update Files in Folder, Deletes Files in Folder and Get Bucket Directory Structure]
- Supported File Formats - JSON, CSV, TSV, AVRO, PARQUET

### CODE REVIEW

**Review by AryaanB9 — APPROVED** [review]

File: internal/tasks/blob_loading/blob_operations.go [hunk 1]
```diff
@@ -0,0 +1,862 @@
...
+
+	var keyValues []external_storage.KeyValue
+	fileFormat := strings.ToLower(extra.FileFormat) // file format can be like "json" or "json, avro" or "json,parquet"
+	if fileFormat == "" {
+		log.Println("create files to insert: file format not provided")
+		return
+	}
+
+	templateName := strings.ToLower(operationConfig.TemplateName)
+	pathToFileOnDisk := TempFolderPath + strconv.FormatInt(result.ResultSeed, 10) + "/" + extra.FilePath
```
Comment (manik-couchbase) [inline internal/tasks/blob_loading/blob_operations.go:204]:
```suggestion
        pathToFileOnDisk := filepath.Join(TempFolderPath,strconv.FormatInt(result.ResultSeed, 10),extra.FilePath)
```

File: internal/tasks/blob_loading/blob_operations.go [hunk 2]
```diff
@@ -0,0 +1,862 @@
...
+			log.Println("create files to insert: create directory:", err)
+			return
+		}
+	} else if err != nil {
+		log.Println("create files to insert: check existence of directory:", err)
+		return
+	}
+
+	// Validating file format
+	if checkFileFormat := external_storage.ValidateFileFormat(fileFormat); checkFileFormat {
```
Comment (manik-couchbase) [inline internal/tasks/blob_loading/blob_operations.go:221]: validateFileFormat() can return (file_format, bool).
If  (returned true) then use the return type file_format
else prepare errors

Handle the file_format extraction and validation in a single function.

File: internal/tasks/blob_loading/blob_operations.go [hunk 3]
```diff
@@ -0,0 +1,862 @@
...
+	defer func(file *os.File) {
+		err := file.Close()
+		if err != nil {
+			log.Println("create files to insert: closing output file:", err)
+			return
+		}
+	}(file)
+
+	// Handling the Different File Formats
+	switch fileFormat {
```
Comment (manik-couchbase) [inline internal/tasks/blob_loading/blob_operations.go:250]: Keep a single outer loop out avoid redundant loops for each file_format.

File: internal/tasks/blob_loading/blob_operations.go [hunk 4]
```diff
@@ -0,0 +1,862 @@
...
```
Comment (manik-couchbase) [inline internal/tasks/blob_loading/blob_operations.go:530]: Same advise to keep the key-value loop as single loop and build file as per the file format.

File: internal/tasks/blob_loading/blob_operations.go (hunk 3 above)
Comment (manik-couchbase) [reply internal/tasks/blob_loading/blob_operations.go:250]: Thanks for clarification

File: internal/tasks/blob_loading/blob_operations.go (hunk 4 above)
Comment (manik-couchbase) [reply internal/tasks/blob_loading/blob_operations.go:530]: Duplicated

File: internal/tasks/blob_loading/blob_loading.go [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+	 * Batch size becomes the number of documents to be inserted to create a single file hence: Batch Size = End - Start
+	 */
+
+	numOfBatches := int64(1)
+	batchSize := task.OperationConfig.End - task.OperationConfig.Start
+	remainingItems := (task.OperationConfig.End - task.OperationConfig.Start) - (numOfBatches * batchSize)
+
+	// If the operation requires the creation of files, then we create the folder for it in TempFolderPath
+	if tasks.CheckFileCreationOperation(task.Operation) {
+		err := os.Mkdir(TempFolderPath+strconv.FormatInt(task.ResultSeed, 10), os.ModePerm)
```
Comment (manik-couchbase) [inline internal/tasks/blob_loading/blob_loading.go:194]: use filepath.Join()

File: internal/tasks/blob_loading/blob_loading.go [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+			if diffMinMax == 0 {
+				// If MaxFileSize == MinFileSize and both are non zero
+				documentSize = task.ExternalStorageExtras.MaxFileSize
+			} else if diffMinMax <= 0 {
+				documentSize = 1048576 // 1 MB
+			} else {
+				documentSize = task.ExternalStorageExtras.MinFileSize + rand.Int63n(diffMinMax)
+			}
+		} else {
+			documentSize = 1048576 // 1 MB
```
Comment (manik-couchbase) [inline internal/tasks/blob_loading/blob_loading.go:233]: make a constant in helper file

File: internal/tasks/blob_loading/blob_loading.go [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+		}
+
+		// Here, we have to update files in a single folder. So we generate the file paths w.r.t that folder
+		//filePaths := GenerateFilePaths([]string{task.ExternalStorageExtras.FolderPath}, task.ExternalStorageExtras.FilesPerFolder, task.ExternalStorageExtras.FileFormat)
+		numOfBatches = int64(len(filePathsToUpdate))
+
+		// Batch Size = Num of Docs to generate to get file size somewhere between MinFileSize and MaxFileSize
+		var documentSize int64
+		if task.ExternalStorageExtras.MaxFileSize > 0 && task.ExternalStorageExtras.MinFileSize > 0 {
+			diffMinMax := task.ExternalStorageExtras.MaxFileSize - task.ExternalStorageExtras.MinFileSize
```
Comment (manik-couchbase) [inline internal/tasks/blob_loading/blob_loading.go:342]: Document size estimation in a single function

File: internal/tasks/blob_loading/blob_loading.go [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+	}
+	t.Result = nil
+
+	t.State.StopStoringState()
+	if err := t.State.SaveTaskSateOnDisk(); err != nil {
+		log.Println("not able to save state into ", t.MetaDataIdentifier(), " ", t.ResultSeed, " ", t.Operation)
+	}
+
+	t.TaskPending = false
+	return t.req.SaveRequestIntoFile()
```
Comment (manik-couchbase) [inline internal/tasks/blob_loading/blob_loading.go:136]:
```suggestion
	defer func() {
		go func() {
			time.Sleep(time.Duration(30 * time.Minute))
			t.Result = nil
		}()
	}()

	defer func() {
		go func() {
			time.Sleep(time.Duration(10 * time.Hour))
			t.State = nil
		}()
	}()
	
		t.Result.StopStoringResult()

	t.State.StopStoringState()

	t.TaskPending = false

	return t.req.SaveRequestIntoFile()
```

**Review by manik-couchbase — COMMENTED** [review]
let me know after you fix the suggestions :D

File: internal/tasks/blob_loading/blob_loading.go (hunk 7 above)
Comment (AryaanB9) [reply internal/tasks/blob_loading/blob_loading.go]: Added a function to calculate size of file

File: internal/tasks/blob_loading/blob_loading.go (hunk 6 above)
Comment (AryaanB9) [reply internal/tasks/blob_loading/blob_loading.go]: Added a constant for default file size

**Pull request merged**